src/GovUk.Education.ExploreEducationStatistics.Data.Processor/Services/DataImportService.cs

```diff
@@ -220,7 +221,112 @@ public async Task WriteDataSetFileMeta(Guid subjectId)
                 .Single(f => f.Type == FileType.Data
                              && f.SubjectId == subjectId);
             file.DataSetFileMeta = dataSetFileMeta;
+
+            file.FilterHierarchies = await GenerateFilterHierarchies(statisticsDbContext, filters);
+
             await contentDbContext.SaveChangesAsync();
         }
+
+        public static async Task<List<DataSetFileFilterHierarchy>> GenerateFilterHierarchies(
+            StatisticsDbContext statisticsDbContext,
+            List<FilterMeta> filters)
+        {
+            var rootFilters = filters
+                .Where(parentFilter => parentFilter.GroupCsvColumn == null
+                                       && filters.Any(childFilter =>
+                                           parentFilter.ColumnName == childFilter.GroupCsvColumn));
+
+            var hierarchies = new List<DataSetFileFilterHierarchy>();
+
+            foreach (var rootFilter in rootFilters)
+            {
+                var hierarchy = await GenerateFilterHierarchy(statisticsDbContext, rootFilter, filters);
+                hierarchies.Add(hierarchy);
+            }
+
+            return hierarchies;
+        }
+
+        private static async Task<DataSetFileFilterHierarchy> GenerateFilterHierarchy(
+            StatisticsDbContext statisticsDbContext,
+            FilterMeta rootFilter,
+            List<FilterMeta> filters)
+        {
+            var rootFilterItemIds = statisticsDbContext.FilterItem
+                .AsNoTracking()
+                .Where(fi => fi.FilterGroup.FilterId == rootFilter.Id)
+                .Select(fi => fi.Id)
+                .ToHashSet();
+
+            var childFilterIds = new List<Guid>();
+            var tiers = new List<Dictionary<Guid, List<Guid>>>();
+
+            var parentFilter = rootFilter;
+            var parentFilterItemIds = rootFilterItemIds;
+            var childFilter = filters
+                .Single(f => f.GroupCsvColumn == parentFilter.ColumnName);
+
+            while (true) // one iteration of loop per tier
```

Comment:
This was actually how I originally had it (pretty much) but changed it thinking this would make it easier to follow the logic at the end of each loop iteration setting up the next.

I'll have a think about how this can be done - cause we've thought there is a slight issue here so probably worth it.

Reply:
I gave this some thought and decided to leave it in the end. I don't think there is a big advantage to shifting this around so better to save the time and just get it in.